service/forward_service.cc

```diff
@@ -412,27 +398,50 @@ future<query::forward_result> forward_service::execute_on_this_shard(
         cql3::query_options::specific_options::DEFAULT,
         cql_serialization_format::latest()
     );
-    auto pager = service::pager::query_pagers::pager(
-        _proxy,
-        schema,
-        selection,
-        *query_state,
-        *query_options,
-        make_lw_shared<query::read_command>(std::move(req.cmd)),
-        retain_ranges_owned_by_this_shard(schema, std::move(req.pr)),
-        nullptr // No filtering restrictions
-    );
+
     auto rs_builder = cql3::selection::result_set_builder(
         *selection,
         now,
         cql_serialization_format::latest(),
         std::vector<size_t>() // Represents empty GROUP BY indices.
     );
 
-    // Execute query.
-    while (!pager->is_exhausted()) {
-        co_await pager->fetch_page(rs_builder, DEFAULT_INTERNAL_PAGING_SIZE, now, timeout);
-    }
+    dht::partition_range_vector ranges_owned_by_this_shard;
+    partition_ranges_owned_by_this_shard owned_iter(schema, std::move(req.pr));
+
+    std::optional<dht::partition_range> current_range;
+    do {
+        // We serve up to 256 ranges at a time to avoid allocating a huge vector for ranges
+        static constexpr size_t max_ranges = 256;
+        while ((current_range = owned_iter.next(*schema))) {
+            ranges_owned_by_this_shard.push_back(*current_range);
+            if (ranges_owned_by_this_shard.size() >= max_ranges) {
+                break;
+            }
+        }
+        if (ranges_owned_by_this_shard.empty()) {
+            break;
```

Comment:
This logic means next() is called even after it returns nullopt. I inspected the partitions_ranges_owned_by_this_shard thing and it tolerates that, but it's bad practice to read past the end. Better to avoid it.

Reply:
I'm pretty sure it was modeled exactly after https://doc.rust-lang.org/std/iter/trait.Iterator.html#tymethod.next from Rust, where next() returns an engaged optional until there's no more data, which is signaled by returning a disengaged one. Wouldn't any other interface expect the caller to always check for has_next()? I can rewrite it, but this semantics look rather natural to me.

Reply:
oh, wait, you meant something else, let me explain: this doesn't mean that next() is called after returning nullopt, on the contrary - it's needed because after reading 256 items we might exit the loop early, but at the same time we just read the last element from the iterator. We don't know it was the last element yet, so we'll enter the loop body again. And in there, we'll get nullopt right after calling next(), which means that we didn't fill ranges_owned_by_this_shard with any elements. And for that corner case, we exit early here in order not to try using a pager on an empty partition ranges vector.

Reply:
You're right. So pending clarification of the other comment, the patch looks good.